maven-artifact/src/main/java/org/apache/maven/artifact/versioning/ComparableVersion.java

```diff
@@ -82,48 +84,181 @@
     }
 
     /**
-     * Represents a numeric item in the version item list.
+     * Represents a numeric item in the version item list that can be represented with an int.
      */
-    private static class IntegerItem
+    private static class IntItem
         implements Item
     {
-        private static final BigInteger BIG_INTEGER_ZERO = new BigInteger( "0" );
+        private final int value;
 
-        private final BigInteger value;
+        public static final IntItem ZERO = new IntItem();
+
+        private IntItem()
+        {
+            this.value = 0;
+        }
+
+        IntItem( String str )
+        {
+            this.value = Integer.parseInt( str );
+        }
+
+        public int getType()
+        {
+            return INT_ITEM;
+        }
+
+        public boolean isNull()
+        {
+            return value == 0;
+        }
+
+        public int compareTo( Item item )
+        {
+            if ( item == null )
+            {
+                return ( value == 0 ) ? 0 : 1; // 1.0 == 1, 1.1 > 1
+            }
+
+            switch ( item.getType() )
+            {
+                case INT_ITEM:
+                {
+                    int itemValue = ( (IntItem) item ).value;
+                    return ( value < itemValue ) ? -1 : ( ( value == itemValue ) ? 0 : 1 );
+                }
+                case LONG_ITEM:
```

Comment:
if `this` is an IntItem and `item` a LongItem and BigIntegerItem, then item is always bigger, right (since we control the optimized value)?

Reply:
Too much of an optimization. Strings are longer but not necessarily numerically bigger. Don't know if this can happen in practice but the following would not be equal versions any more. With that optimization the results are wrong:
java -jar target/maven-artifact-3.6.1-SNAPSHOT.jar 1  0000000001
1. 1 == 1
   1 < 0000000001
2. 0000000001 == 1
java -jar target/maven-artifact-3.6.1-SNAPSHOT.jar 1  000000000000000000001
1. 1 == 1
   1 < 000000000000000000001
2. 000000000000000000001 == 1

Reply:
How about stripping the leftside zero's before parsing? And these are good cases to write a unittest for.